src/TgSharp.Core/Network/WSTransport.cs

```diff
@@ -0,0 +1,178 @@
+                {
+                    await ReconnectionHappened(msg);
+                }))
+                .Concat()
+                .Subscribe();
+
+            await client.StartOrFail();
+        }
+
+        private async Task ReconnectionHappened(ReconnectionInfo obj)
+        {
+            if (client.IsRunning)
+                await InitializeConnection();
+        }
+
+        private async Task HandleIncomingMessage(ResponseMessage msg)
+        {
+            byte[] unencryptedData = new byte[msg.Binary.Length];
+            AesCtr.Ctr128Encrypt(msg.Binary, decryptKey, ref decryptIV, ref decryptCount, ref decryptNum, unencryptedData);
+
+            var tcpMessage = Message.Decode(unencryptedData);
+
+            var authKeyId = BitConverter.ToInt64(tcpMessage.Body, 0);
+
+            if (authKeyId == 0)
+                OnUnencryptedMessage?.Invoke(tcpMessage);
+            else
+                OnEncryptedMessage?.Invoke(tcpMessage);
+        }
+
+        private async Task InitializeConnection()
+        {
+            byte[] init = new byte[64];
+
+            while (true)
+            {
+                rngSource.GetNonZeroBytes(init);
+                if (init[0] == 0xef)
+                    continue;
+
+                uint firstInt = BitConverter.ToUInt32(init, 0);
+                if (firstInt == 0x44414548 || firstInt == 0x54534f50 || firstInt == 0x20544547 || firstInt == 0x4954504f || firstInt == 0x02010316 || firstInt == 0xdddddddd || firstInt == 0xeeeeeeee)
+                    continue;
+
+                uint secondInt = BitConverter.ToUInt32(init, 4);
+                if (secondInt == 0x00000000)
+                    continue;
+
+                break;
+            }
+
+            Buffer.BlockCopy(BitConverter.GetBytes(protocol), 0, init, 56, 4);
+
+            byte[] initRev = init.Reverse().ToArray();
+
+            encryptKey = new byte[32];
+            Buffer.BlockCopy(init, 8, encryptKey, 0, 32);
+            decryptKey = new byte[32];
+            Buffer.BlockCopy(initRev, 8, decryptKey, 0, 32);
+
+            encryptIV = new byte[16];
+            Buffer.BlockCopy(init, 40, encryptIV, 0, 16);
+            decryptIV = new byte[16];
+            Buffer.BlockCopy(initRev, 40, decryptIV, 0, 16);
+
+            byte[] encryptedInit = new byte[init.Length];
+            AesCtr.Ctr128Encrypt(init, encryptKey, ref encryptIV, ref encryptCount, ref encryptNum, encryptedInit);
+
+            byte[] finalInit = new byte[64];
+            Buffer.BlockCopy(init, 0, finalInit, 0, 56);
+            Buffer.BlockCopy(encryptedInit, 56, finalInit, 56, 8);
+
+            await client.SendInstant(finalInit);
+        }
+
+        public async Task ConnectAsync(int retryCount = 5)
+        {
+            while (retryCount-- != 0)
+            {
+                try
+                {
+                    await InitializeClient();
+                    break;
+                }
+                catch (Exception ex)
```

Comment:
what kind of exception can happen here? let's put the specific type, not generic Exception base class

Reply:
not sure but the point is that in case of any error I want to retry there are a lot of socket related exceptions

Reply:
can you gist 2 exceptions of this, and make retryCount default to 0